src/generate-erc721-permit-signature.ts

```diff
@@ -2,12 +2,25 @@ import { getPayoutConfigByNetworkId } from "./utils/payoutConfigByNetworkId";
 import { BigNumber, ethers, utils } from "ethers";
 import { MaxUint256 } from "@uniswap/permit2-sdk";
 import { keccak256, toUtf8Bytes } from "ethers/lib/utils";
+import { Erc721PermitSignatureData, GenerateErc721PermitSignatureParams, PermitTransactionData } from "./types/permits";
 
-const NFT_MINTER_PRIVATE_KEY = process.env.NFT_MINTER_PRIVATE_KEY ?? "0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d";
-const NFT_CONTRACT_ADDRESS = "0x38a70c040ca5f5439ad52d0e821063b0ec0b52b6"; // "0x6a87f05a74AB2EC25D1Eea0a3Cd24C3A2eCfF3E0";
+const NFT_MINTER_PRIVATE_KEY = process.env.NFT_MINTER_PRIVATE_KEY;
+const NFT_CONTRACT_ADDRESS = process.env.NFT_CONTRACT_ADDRESS;
```

Comment:
You should immediately check after if it exists. Generally we use helper functions to 1. import all expected `process.env` values, and 2. check if they exist.

Reply:
Also I just realized but in a normal node runtime you must use dotenv in every file or else this shouldn't work. Did you test your deliverable? If its not tested this is not ready for review.

Reply:
- https://github.com/ubiquity/pay.ubq.fi/pull/177#issue-2152854389 

I posted a test video showing it's all functioning however not within a node env, I'm setting that up now first for ``/research`` then ``/permit``. 

My current setup looks like: 

kernel -> pluginDispatch() -> plugin workflow -> runs the action -> posts the comment

> I thought that plugins will use repository dispatch to return data to the kernel, which will call the next plugin in the chain or stop if it's the last one.

My setup requires that no data be returned to the kernel as my plugins would handle the entire execution flow after dispatch, not sure if that's what you intend for it but it will suffice for testing

Reply:
> My setup requires that no data be returned to the kernel as my plugins would handle the entire execution flow after dispatch, not sure if that's what you intend for it but it will suffice for testing

Sure for early testing it might be fine. @whilefoo please chime in on how the interface will look like for returning results from plugins. 

---

The plugins must return results to the kernel because the kernel is what reads the config and connects the chain of plugins together. 

Each partner instance can configure their own order of plugins. 

This is only sensible to implement the logic once inside the kernel to dynamically handle mounting multiple plugins in whatever order.

Reply:
Where I'm confused is: invoking an action.yml is different from a workflow, action.yml cannot be executed on a dispatch unless called from a workflow which would imply each plugin will require a workflow listening for it's custom event and the kernel would kick off the workflow which would invoke the action.yml otherwise won't the kernel require some kind of robust generic workflow that it can begin that invokes these action.yml's?

The way I have it set up now, the config.yml is parsed, all plugins for the new comment event  are initiated in plugins, it maps over the regex's applying them then firing the plugin (starting the workflow in the plugin's repo which handles sequencing of the other plugins it relies on by the way of the uses keyword) if it matches any.

If the kernel is piping output from one workflow into the next that's going to take ages running through setup 2/3/4 times as opposed one workflow orchestrating the sequence of other plugins/core actions that it relies on. It's still very modular, partner-configurable and would be quicker than waiting for multiple runners to be assigned, built and executed.

Maybe I'm off on the wrong track, idk but I can't wrap my head around it from a plugin dev standpoint. Partners just add the plugin's info into their config.yml obtained from the plugin's README, the dev defines the config and sequence of called actions and if the kernel is using dispatch to kickoff multiple sequencial workflows then it could just call one workflow that encapsulates everything that plugin needs and then the kernel can be hands-off immediately after dispatch

As custom mounting etc is to be avoided without hindering plugin extensibility I thought it made sense that the kernel scans the config for plugins, applies those regexes on the appropriate webhook events then allowing the plugin's workflow to handle everything.

Are their more reasons to send data back to the kernel other than 'piping'?

Reply:
It seems that your primarily concern with this architecture is speed. 

As per my original proposal we can easily compile popular recipes into something like a cloudflare worker if speed is a concern. 

For example the recipe for conversation rewards by default would be three separate GitHub actions in my proposal. But because we use it so frequently it might make sense to compile into a single Worker and use that endpoint as the "workflow" (in this case it would make sense to paste in the entire URL to indicate to the bot that it's an external API)

--- 

Beyond this, it makes sense to conform to GitHub's existing Action schema to take advantage of all of the actions that already exist. We can make the bot interoperate seamlessly with all of those existing capabilities. 

For example, with my proposal it would be trivial to map any action to any arbitrary slash command. The possibilities are infinite though. 

It makes our offering more capable without much extra work. 

---

Lastly it technically would allow us to directly connect our plugins into CI if we wanted to. 

---

I don't understand your concern with action.yml. It can be a simple wrapper that just calls an action inside of the workflows folder. It's a non issue.

Reply:
Thanks for this I'm sure I'm on the same page now

Convo Rewards modules:
- commentScoring (action.yml)
- permitGeneration (action.yml)
- permitComment (action.yml)

This recipe is invoked via dispatch (starts a workflow in the recipe repo otherwise it would need to kick off one from the kernel/bot repo which I don't expect) with ``uses`` for each module sequenced as per the workflow file (``ubiquibot-config.plugins?.ISSUE_COMMENT_CREATED`` would match this sequence adhering to the plugin config structure you described

Each module is a lego brick that a workflow imports and runs with or without other modules

If a recipe takes an exorbitant amount of time a worker would be built for that recipe

> I don't understand your concern with action.yml. It can be a simple wrapper that just calls an action inside of the workflows folder. It's a non issue.

Wouldn't it be the other way about, the ``/workflows/*.yml`` calls the re-usable ``action.yml``s and then if piping the output from a workflow (which has run the imported actions) you'd access that through the ``actions/runs/${runID}/jobs`` endpoint?

From what I've read and in building what I have I've found out that dispatching a workflow does not return a runID, you'd need to fetch from ``/actions/runs/`` and filter for ``in_progress`` then match a runner to the workflow that was just dispatched (assuming heavy traffic it might be difficult to find the relevant run). Will know for sure later today

Reply:
> Wouldn't it be the other way about, the /workflows/*.yml calls the re-usable action.yml

Don't see why that would be the case. Perhaps there is a gap in my knowledge, but from what I understand `action.yml` is the default entry point for any GitHub Action, so we should follow this. 

Also as a side note, I realize that `conversation-rewards` (the first step of the recipe, which includes qualitative and quantitative analysis of comments) would be impossible to include in a Worker. It was actually the reason why we invented the "delegated compute" infrastructure in the first place (being able to run event handlers outside of the bot kernel.) 

It takes a tremendous amount of time to render comments in a virtual DOM (this is to be able to consolidate both markdown and HTML into a single HTML selector for `ubiquibot-config.yml` purposes) as well as to run the Open AI queries for qualitative analysis. From memory it took ~400ms per comment for quantitative analysis - even running locally on an M1 Macbook - with average issues having around 10 comments to process including their pull requests. And then several network requests to OpenAI, which returns its results over a couple of seconds. 

Workers free tier [has a max of 10ms](https://developers.cloudflare.com/workers/platform/limits/#worker-limits) for perspective.

Reply:
Seeing whilefoo's recent push I have a better understanding of how things are going to be running

comment created webhook fires, which I assume uses ``workflow_dispatch`` to start the plugin and the plugin is expected to return the output with a ``repo_dispatch`` dispatch step (I think)

``/research`` needs to be able to read from the botConfig for the openai key, and I see that the authToken is being passed as an input (which I think is insecure if that authToken can read from private repos idk) via ``DelegatedComputeInputs`` I haven't fired any workflows this way yet as I don't want to log anything sensitive(unsure if it would). On my build and with this one that's where I'm stuck.

So I was thinking of setting repo secrets (assuming that's how the @ubiquibot/plugin repos will work too) which in the context of the ``/research-plugin`` repo where the action will run it would have that private read access without sending the app install auth token?

Should my research action be expecting ``npx tsx "body" issueNumber sender org``

or should it be expecting to handle the original event and destructuring everything it needs (token, body, etc..)